### PR TITLE
Changing tensorflow implementation in RecoHGCal/TICL

### PR DESCRIPTION
**Pull request by dr-stringfellow**
#### PR description:

The tensorflow model used to regress energy and ID of TICL objects ("tracksters") has been moved from an ED-based implementation to an ES setup. All plugins (CA, CLUE3D, FastJet, ...) will now access the same cached model, instead of caching one model for each plugin as was the case in the EDProducer setup. 

This is now in line with the RecoTracker tensorflow implementation. In this first step, we are pointing also to their tensorflow producer. Eventually, in a pull request that would follow soon, this should be harmonised and moved to a central, neutral place that both RecoTracker and RecoHGCal use.

#### PR validation:

It saves about 5% in memory on upgrade workflow
`39034.202`:

https://bmaier.web.cern.ch/bmaier/view.php?dir=figs/hgcal/ticltf/new

(The more plugins are run in the same job, the higher the savings.)

### CODE REVIEW

**Comment by rovere** [comment]
Thanks a lot, @dr-stringfellow for this nice piece of work!

**Comment by rovere** [comment]
@cmsbuild please test

**Comment by rovere** [comment]
@jpata @clacaputo would you prefer to enable also the tests with profiling enabled, or are you satisfied with the plots and measurements that Benedikt already did and linked in the description of this PR?

**Comment by jpata** [comment]
@cmsbuild enable profiling

let's run with profiling, why not.

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
Note that the static checker complains about the const_cast (also an issue for the trackDNN, related to https://github.com/cms-sw/cmssw/issues/32529. It's not something that can be fixed in this PR, but to be kept in mind.
```
4 warnings generated.
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2022-01-18-2300/src/RecoHGCal/TICL/plugins/PatternRecognitionbyFastJet.cc:272:19: warning: const_cast was used, this may result in thread-unsafe code [threadsafety.ConstCast]
  tensorflow::run(const_cast<tensorflow::Session *>(eidSession), inputList, outputNames, &outputs);
                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2022-01-18-2300/src/RecoHGCal/TICL/plugins/PatternRecognitionbyCA.cc:439:19: warning: const_cast was used, this may result in thread-unsafe code [threadsafety.ConstCast]
  tensorflow::run(const_cast<tensorflow::Session *>(eidSession), inputList, outputNames, &outputs);
                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2022-01-18-2300/src/RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc:662:19: warning: const_cast was used, this may result in thread-unsafe code [threadsafety.ConstCast]
  tensorflow::run(const_cast<tensorflow::Session *>(eidSession), inputList, outputNames, &outputs);
                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2022-01-18-2300/src/RecoHGCal/TICL/plugins/PatternRecognitionbyCLUE3D.cc:403:19: warning: const_cast was used, this may result in thread-unsafe code [threadsafety.ConstCast]
  tensorflow::run(const_cast<tensorflow::Session *>(eidSession), inputList, outputNames, &outputs);
                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I agree that in a follow-up (not necessarily by the HGCAL team), the `TfGraphDefProducer` stuff should be moved out of RecoTracker to a common area.

**Comment by jpata** [comment]
For posterity, can you comment on what settings were used to make this plot (besides the worfklow 39034.202):
image

**Comment by jpata** [comment]
enable profiling

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by dr-stringfellow** [comment]
> For posterity, can you comment on what settings were used to make this plot (besides the worfklow 39034.202):
> image

The following settings were used:

`cmsRun step3 -s RAW2DIGI,L1Reco,RECO,RECOSIM,PAT,VALIDATION:@phase2Validation+@miniAODValidation,DQM:@phase2+@miniAODDQM --conditions auto:phase2_realistic_T27 --datatier GEN-SIM-RECO,MINIAODSIM,DQMIO -n 10 --eventcontent FEVTDEBUGHLT,MINIAODSIM,DQM --geometry Extended2026D87 --era Phase2C11I13T27M9 --procModifiers fastJetTICL,clue3D --nThreads 1 --customise Validation/Performance/TimeMemoryInfo.py`

**Comment by jpata** [comment]
https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_12_3_X_2022-01-24-2300/slc7_amd64_gcc10/profiling/23434.21/igprofMEM_step3.19
Screenshot from 2022-01-26 11-48-05

https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_12_3_X_2022-01-24-2300/slc7_amd64_gcc10/profiling/23434.21/PR-95bd10/21972/igprofMEM_step3.19
Screenshot from 2022-01-26 11-48-21

we also see the improvement in the bot igprof job.

**Comment by jpata** [comment]
+reconstruction
- technical, improves the memory handling of the TICL TF models by loading them from the EventSetup
- no reco changes expected/observed

**Comment by AdrianoDee** [comment]
+upgrade
Technical, improving memory for upgrade wfs including TICL.

File: RecoHGCal/TICL/plugins/PatternRecognitionbyCA.cc [hunk 1]
```diff
@@ -446,7 +436,7 @@ void PatternRecognitionbyCA<TILES>::energyRegressionAndID(const std::vector<reco
   }
 
   // run the inference (7)
-  tensorflow::run(eidSession_, inputList, outputNames, &outputs);
+  tensorflow::run(const_cast<tensorflow::Session *>(eidSession), inputList, outputNames, &outputs);
```
Comment (perrotta) [inline RecoHGCal/TICL/plugins/PatternRecognitionbyCA.cc:439]: `eidSession` is already a `const tensorflow::Session *`: why do you need the `const_cast`?

File: RecoHGCal/TICL/plugins/PatternRecognitionbyCLUE3D.cc [hunk 2]
```diff
@@ -410,7 +400,7 @@ void PatternRecognitionbyCLUE3D<TILES>::energyRegressionAndID(const std::vector<
   }
 
   // run the inference (7)
-  tensorflow::run(eidSession_, inputList, outputNames, &outputs);
+  tensorflow::run(const_cast<tensorflow::Session *>(eidSession), inputList, outputNames, &outputs);
```
Comment (perrotta) [inline RecoHGCal/TICL/plugins/PatternRecognitionbyCLUE3D.cc:403]: `eidSession` is already a `const tensorflow::Session *`: why do you need the `const_cast`?

File: RecoHGCal/TICL/plugins/PatternRecognitionbyFastJet.cc [hunk 3]
```diff
@@ -278,7 +269,7 @@ void PatternRecognitionbyFastJet<TILES>::energyRegressionAndID(const std::vector
   }
 
   // run the inference (7)
-  tensorflow::run(eidSession_, inputList, outputNames, &outputs);
+  tensorflow::run(const_cast<tensorflow::Session *>(eidSession), inputList, outputNames, &outputs);
```
Comment (perrotta) [inline RecoHGCal/TICL/plugins/PatternRecognitionbyFastJet.cc:272]: `eidSession` is already a `const tensorflow::Session *`: why do you need the `const_cast`?

File: RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc [hunk 4]
```diff
@@ -655,7 +659,7 @@ void TrackstersMergeProducer::energyRegressionAndID(const std::vector<reco::Calo
   }
 
   // run the inference (7)
-  tensorflow::run(eidSession_, inputList, outputNames, &outputs);
+  tensorflow::run(const_cast<tensorflow::Session *>(eidSession), inputList, outputNames, &outputs);
```
Comment (perrotta) [inline RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc:662]: `eidSession` is already a `const tensorflow::Session *`: why do you need the `const_cast`?

File: RecoHGCal/TICL/plugins/PatternRecognitionbyCA.cc (hunk 1 above)
Comment (jpata) [reply RecoHGCal/TICL/plugins/PatternRecognitionbyCA.cc:439]: this is removing the const (but tensorflow::run is known to be thread-safe), to fit the the tensorflow interface.
related to https://github.com/cms-sw/cmssw/issues/35541

**Comment by perrotta** [comment]
+1

**Pull request merged**